lib/datasource/datasource.ts

```diff
@@ -0,0 +1,51 @@
+import { ExternalHostError } from '../types/errors/external-host-error';
+import { Http } from '../util/http';
+import type { HttpError } from '../util/http/types';
+import type {
+  DatasourceApi,
+  DigestConfig,
+  GetReleasesConfig,
+  ReleaseResult,
+} from './types';
+
+export abstract class Datasource implements DatasourceApi {
+  protected constructor(public readonly id: string) {
+    this.http = new Http(id);
+  }
+
+  caching: boolean;
+
+  customRegistrySupport = true;
+
+  defaultConfig: Record<string, unknown>;
+
+  defaultRegistryUrls: string[];
+
+  defaultVersioning: string;
+
+  registryStrategy: 'first' | 'hunt' | 'merge' = 'first';
```

Comment:
```suggestion
  caching: boolean | undefined;

  customRegistrySupport = true;

  defaultConfig: Record<string, unknown> | undefined;

  defaultRegistryUrls: string[] | undefined;

  defaultVersioning: string | undefined;

  registryStrategy: 'first' | 'hunt' | 'merge' | undefined = 'first';
```

They should stay optional and we want to make future prove for typescript `strictNullCheck`.